src/server/acl/acl_family.cc

```diff
@@ -247,12 +247,25 @@ std::optional<facade::ErrorReply> AclFamily::LoadToRegistryFromFile(std::string_
     requests.push_back(std::move(std::get<User::UpdateRequest>(req)));
   }
 
+  absl::flat_hash_set<std::string_view> usernames_set;
+  for (const auto& username : usernames) {
+    usernames_set.insert(username);
+  }
+
   auto registry_with_wlock = registry_->GetRegistryWithWriteLock();
   auto& registry = registry_with_wlock.registry;
   // TODO(see what redis is doing here)
   if (!init) {
     registry.clear();
   }
+  // Evict open connections for users not present in the updated ACL
+  for (const auto& [username, _] : registry) {
+    if (username == "default")
+      continue;
+    if (!usernames_set.contains(username)) {
+      EvictOpenConnectionsOnAllProactors(username);
+    }
+  }
```

Comment:
The second `for-loop` won't work because the registry is `cleared` on line `258`. 

Furthermore, I don't like that we call `EvictOpenConnectionsOnAllProactors` for each `username`. We can do that in one pass for all the connections:

Introduce a new function, `EvictOpenConnectionsOnAllProactorsWithUsers` and add an extra argument `absl::flat_hash_set` which is then captured in the `close_cb`. Then replace `ctx->authed_username == user`, with `contains`. That way we call this function once:

`EvictOpenConnectionsOnAllProactors(usernames_set)`. (Bonus points if you can merge the functionality with the current function -- not mandatory though :stuck_out_tongue: )

Furthermore, I realized that `redis` is evicting all open connections even when there are common usernames.

I would suggest to:

1. remove the `username_set`, and use the newly function directly `EvictOpenConnectionsOnAllProactors(registry)`
2. remove lines 283 - 285 since all open connections are evicted

Reply:
Thank you for the detailed guidance. 
Just to clarify, `Registry` is a map, not a set. Here are three potential solutions, please let me know which one you prefer. I updated my PR by option 2.

1. An overloaded `EvictOpenConnectionsOnAllProactors(const absl::flat_hash_set<std::string_view>& users_set)`.
    Generate a `username_set` from register before calling.
2. `EvictOpenConnectionsOnAllProactorsWithRegistry(const UserRegistry::RegistryType& registry)`.
3. `EvictOpenConnectionsOnAllProactorsWithUsers(const absl::flat_hash_map<std::string_view, User>& users_map)`. 

Tested `./dragonfly --aclfile ./test_acl --alsologtostderr` in GDB on the server side. 
Since we close all the connections, `ACL LOAD` prints an "Error" to the client, which can be optimized.  
_127.0.0.1:6379> ACL LOAD
Error: Server closed the connection
127.0.0.1:6379> ACL list
1) user peter off nopass +@NONE
2) user default on nopass +@ALL +ALL
3) user amy off nopass +@NONE_

I can also add unit test for this use case if you could give me some adivce.

Reply:
Hi @azuredream, thank you so much for the changes! They are :fire: 

I like the (2) approach! 

For what you ask:
> _127.0.0.1:6379> ACL LOAD Error: Server closed the connection

This is good and on par with redis! 

Maybe a small nit, is to `exclude` default, since this user can't be removed by the system and therefore, there is no point to evict them. So maybe, add an extra check:

```
ctx->authed_username != "default" && registry.contains(ctx->authed_username)  
```

Reply:
p.s. no need to do add anything else, I will approve :) 

p.s.2 if you fancy making changes and learning the codebase, you can add a test in `acl_family.py` as well but I leave that up to you!